### PR TITLE
Integrate group_index with local_group_index

### PR DESCRIPTION
**Pull request by megrogan**

### CODE REVIEW

File: backend/canisters/group_index/impl/Cargo.toml [hunk 1]
```diff
@@ -35,4 +35,6 @@ user_canister = { path = "../../user/api" }
 user_canister_c2c_client = { path = "../../user/c2c_client" }
 user_index_canister = { path = "../../user_index/api" }
 user_index_canister_c2c_client = { path = "../../user_index/c2c_client" }
+local_group_index_canister = { path = "../../local_group_index/api" }
```
Comment (hpeebles) [inline backend/canisters/group_index/impl/Cargo.toml:38]: The dependencies are always in alphabetical order

File: backend/canisters/group_index/impl/src/lifecycle/post_upgrade.rs [hunk 2]
```diff
@@ -1,11 +1,15 @@
+use std::time::Duration;
+
```
Comment (hpeebles) [inline backend/canisters/group_index/impl/src/lifecycle/post_upgrade.rs:2]: Erroneous whitespace

File: backend/canisters/group_index/impl/src/lifecycle/post_upgrade.rs [hunk 3]
```diff
@@ -36,9 +40,51 @@ fn post_upgrade(args: Args) {
...
+        let public_groups = state.data.public_groups.iter().map(|g| Group {
+            chat_id: g.id(),
+            wasm_version: g.wasm_version(),
+        });
+        private_groups.chain(public_groups).collect()
+    });
+
+    let group_ids: Vec<_> = groups.iter().map(|g| g.chat_id).collect();
+
+    let index_id: Principal = Principal::from_text("first_local_group_index_id").unwrap();
```
Comment (hpeebles) [inline backend/canisters/group_index/impl/src/lifecycle/post_upgrade.rs:66]: TODO

File: backend/canisters/group_index/impl/src/model/local_group_index_map.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+pub struct LocalGroupIndex {
+    pub group_count: u32,
+    pub full: bool,
+    pub cycle_top_ups: Vec<CyclesTopUp>,
+    pub wasm_version: Version,
+}
+
+impl LocalGroupIndexMap {
+    pub fn add_index(&mut self, index_id: CanisterId) -> bool {
+        self.index_map.insert(index_id, LocalGroupIndex::default()).is_none()
```
[the remark this answers is not in the document]
Comment (hpeebles) [reply backend/canisters/group_index/impl/src/model/local_group_index_map.rs:22]: You should use `try_insert`

File: backend/canisters/group_index/impl/src/updates/set_max_concurrent_group_canister_upgrades.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    let args = local_group_index_canister::c2c_set_max_concurrent_group_upgrades::Args {
+        value: args.value as u32,
+    };
+
+    let futures: Vec<_> = canisters
+        .into_iter()
+        .map(|canister_id| local_group_index_canister_c2c_client::c2c_set_max_concurrent_group_upgrades(canister_id, &args))
+        .collect();
+
+    futures::future::join_all(futures).await;
```
Comment (hpeebles) [inline backend/canisters/group_index/impl/src/updates/set_max_concurrent_group_canister_upgrades.rs:23]: We should check that they all actually succeeded.

That could be a follow on PR though, but we should label it as a TODO.

File: backend/canisters/group_index/impl/src/updates/upgrade_group_canister_wasm.rs [hunk 6]
```diff
@@ -1,45 +1,44 @@
...
+    let args = local_group_index_canister::c2c_upgrade_group_canister_wasm::Args {
+        group_canister_wasm: args.group_canister_wasm.clone(),
+    };
+
+    let futures: Vec<_> = local_group_index_canisters
+        .into_iter()
+        .map(|canister_id| local_group_index_canister_c2c_client::c2c_upgrade_group_canister_wasm(canister_id, &args))
+        .collect();
+
+    futures::future::join_all(futures).await;
```
Comment (hpeebles) [inline backend/canisters/group_index/impl/src/updates/upgrade_group_canister_wasm.rs:28]: Same here, we should ensure they all succeeded.

File: backend/canisters/group_index/impl/src/model/local_group_index_map.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (megrogan) [reply backend/canisters/group_index/impl/src/model/local_group_index_map.rs:22]: try_insert is a nightly-only experimental API

**Review by hpeebles — APPROVED** [review]

**Pull request merged**